storage/src/nodestore/alloc.rs

```diff
@@ -147,12 +148,158 @@ pub fn area_size_to_index(n: u64) -> Result<AreaIndex, Error> {
         })
 }
 
-/// Objects cannot be stored at the zero address, so a [`LinearAddress`] is guaranteed not
-/// to be zero. This reserved zero can be used as a [None] value for some use cases. In particular,
-/// branches can use `Option<LinearAddress>` which is the same size as a [`LinearAddress`]
-pub type LinearAddress = NonZeroU64;
+// pub type LinearAddress = NonZeroU64;
+/// A linear address in the nodestore storage.
+///
+/// This represents a non-zero address in the linear storage space.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[repr(transparent)]
+pub struct LinearAddress(NonZeroU64);
+
+#[expect(unsafe_code)]
+unsafe impl bytemuck::ZeroableInOption for LinearAddress {}
+#[expect(unsafe_code)]
+unsafe impl bytemuck::PodInOption for LinearAddress {}
+
+impl LinearAddress {
+    /// Create a new `LinearAddress`, returns None if value is zero.
+    #[expect(clippy::missing_panics_doc)]
+    #[inline]
+    #[must_use]
+    pub const fn new(addr: u64) -> Option<Self> {
+        match addr {
+            0 => None,
+            _ => Some(LinearAddress(NonZeroU64::new(addr).expect("non zero"))),
+        }
+    }
```

Comment:
expect is unnecessary here:

```suggestion
    pub const fn new(addr: u64) -> Option<Self> {
        match NonZeroU64::new(addr) {
            Some(addr) => Some(LinearAddress(addr)),
            None => None,
        }
    }
```